### PR TITLE
Nh/fix 1884 listener trigger

### PR DESCRIPTION
**Pull request by nhachicha**
register RealmObject created with + add their UnitTest
- [x] copyOrUpdate
- [x] createAllFromJson
- [x] createObjectFromJson
- [x] createOrUpdateAllFromJson
- [x] createOrUpdateObjectFromJson
- [x] add more unit test to cover callback trigger for the above methods

### CODE REVIEW

**Comment by nhachicha** [comment]
@realm/java

File: realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java [hunk 1]
```diff
@@ -67,6 +79,442 @@ public void run() {
...
+
+                assertEquals(primaryKeyAsLong, primaryKeyAsLong2);
+            }
+        });
+
+        TestHelper.awaitOrFail(signalTestFinished);
+        assertEquals(1, typebasedCommitInvocations.get());
+    }
+
+    //UC 0 using Realm.copyToRealmOrUpdate
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java]: Realm.createObjectFromJson

File: realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java [hunk 2]
```diff
@@ -67,6 +79,442 @@ public void run() {
...
+                    fail(e.getMessage());
+                }
+            }
+        });
+
+        TestHelper.awaitOrFail(signalTestFinished);
+        assertEquals(1, typebasedCommitInvocations.get());
+    }
+
+    //UC 0 using Realm.copyToRealmOrUpdate
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java]: Realm.createObjectFromJson

File: realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java [hunk 3]
```diff
@@ -67,6 +79,442 @@ public void run() {
...
+                    fail(e.getMessage());
+                }
+            }
+        });
+
+        TestHelper.awaitOrFail(signalTestFinished);
+        assertEquals(1, typebasedCommitInvocations.get());
+    }
+
+    //UC 0 using Realm.createOrUpdateObjectFromJson
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java]: createOrUpdateObjectFromJson (inputstream)

File: realm/realm-library/src/main/java/io/realm/DynamicRealm.java [hunk 4]
```diff
@@ -74,7 +74,11 @@ public DynamicRealmObject createObject(String className) {
         checkIfValid();
         Table table = schema.getTable(className);
         long rowIndex = table.addEmptyRow();
-        return get(DynamicRealmObject.class, className, rowIndex);
+        DynamicRealmObject dynamicRealmObject = get(DynamicRealmObject.class, className, rowIndex);
+        if (handlerController != null) {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/DynamicRealm.java]: This is also part of the above `get` method?

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 5]
```diff
@@ -473,7 +473,11 @@ private static void initializeRealm(Realm realm) {
         }
 
         try {
-            return configuration.getSchemaMediator().createOrUpdateUsingJsonObject(clazz, this, json, false);
+            E realmObject = configuration.getSchemaMediator().createOrUpdateUsingJsonObject(clazz, this, json, false);
+            if (handlerController != null) {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: Is this really needed. I would assume it uses `get` under the hood?

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 6]
```diff
@@ -496,7 +500,11 @@ private static void initializeRealm(Realm realm) {
         }
         checkHasPrimaryKey(clazz);
         try {
-            return configuration.getSchemaMediator().createOrUpdateUsingJsonObject(clazz, this, json, true);
+            E realmObject = configuration.getSchemaMediator().createOrUpdateUsingJsonObject(clazz, this, json, true);
+            if (handlerController != null) {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: Same

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 7]
```diff
@@ -589,11 +599,15 @@ private static void initializeRealm(Realm realm) {
         } else {
             JsonReader reader = new JsonReader(new InputStreamReader(inputStream, "UTF-8"));
             try {
-                return configuration.getSchemaMediator().createUsingJsonStream(clazz, this, reader);
+                realmObject = configuration.getSchemaMediator().createUsingJsonStream(clazz, this, reader);
             } finally {
                 reader.close();
             }
         }
+        if (realmObject != null && handlerController != null) {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: Same

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 8]
```diff
@@ -620,7 +634,11 @@ private static void initializeRealm(Realm realm) {
         try {
             scanner = getFullStringScanner(in);
             JSONObject json = new JSONObject(scanner.next());
-            return configuration.getSchemaMediator().createOrUpdateUsingJsonObject(clazz, this, json, true);
+            E realmObject = configuration.getSchemaMediator().createOrUpdateUsingJsonObject(clazz, this, json, true);
+            if (handlerController != null) {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: Same

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 9]
```diff
@@ -679,7 +698,11 @@ void remove(Class<? extends RealmObject> clazz, long objectIndex) {
      */
     public <E extends RealmObject> E copyToRealm(E object) {
         checkNotNullObject(object);
-        return copyOrUpdate(object, false);
+        E realmObject = copyOrUpdate(object, false);
+        if (handlerController != null) {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: Same

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 10]
```diff
@@ -695,7 +718,11 @@ void remove(Class<? extends RealmObject> clazz, long objectIndex) {
     public <E extends RealmObject> E copyToRealmOrUpdate(E object) {
         checkNotNullObject(object);
         checkHasPrimaryKey(object.getClass());
-        return copyOrUpdate(object, true);
+        E realmObject = copyOrUpdate(object, true);
+        if (handlerController != null) {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: Same

File: changelog.txt [hunk 11]
```diff
@@ -6,6 +6,7 @@
  * Updated ProGuard configuration (#1904).
  * Fixed a bug which could cause a crash when closing Realm instances in change listeners (#1900).
  * Fix a crash occurring during update of multiple async queries (#1895).
+ * Fix (#1884) listeners not triggered for RealmObject & RealmResults created using copy or create methods.
```
Comment (cmelchior) [inline changelog.txt]: We have the issue number at the end for all other entries.

**Comment by cmelchior** [comment]
:+1: 

@realm/java

File: realm/realm-library/src/androidTest/java/io/realm/TestHelper.java [hunk 12]
```diff
@@ -677,4 +677,9 @@ public static void exitOrThrow(final ExecutorService executorService,
             throw throwable[0];
         }
     }
+
```
Comment (kneth) [inline realm/realm-library/src/androidTest/java/io/realm/TestHelper.java]: Thanks for moving it!

File: realm/realm-library/src/main/java/io/realm/HandlerController.java [hunk 13]
```diff
@@ -581,4 +581,36 @@ private void deleteWeakReferences() {
...
+
+    void addToRealmResults(RealmResults<? extends RealmObject> realmResults) {
+        WeakReference<RealmResults<? extends RealmObject>> realmResultsWeakReference
+                = new WeakReference<RealmResults<? extends RealmObject>>(realmResults, referenceQueueSyncRealmResults);
+        syncRealmResults.add(realmResultsWeakReference);
+    }
+
+    // add to the list of RealmObject to be notified after a commit
+    <E extends RealmObject> void addToRealmObjects(E realmobject) {
+            realmObjects.put(new WeakReference<RealmObject>(realmobject), null);
```
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/HandlerController.java]: wrong indent?

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 14]
```diff
@@ -589,11 +596,12 @@ private static void initializeRealm(Realm realm) {
         } else {
             JsonReader reader = new JsonReader(new InputStreamReader(inputStream, "UTF-8"));
             try {
-                return configuration.getSchemaMediator().createUsingJsonStream(clazz, this, reader);
+                realmObject = configuration.getSchemaMediator().createUsingJsonStream(clazz, this, reader);
             } finally {
                 reader.close();
             }
         }
+        return realmObject;
```
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: missing `addToRealmObjects`?

Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: No, because all the `addCollection` methods return void

**Comment by kneth** [comment]
Looks fine to me.

**Comment by zaki50** [comment]
:+1: apart from minor indent issue

File: realm/realm-library/src/main/java/io/realm/HandlerController.java (hunk 13 above)
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/HandlerController.java]: I'll update this

Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/HandlerController.java]: pushed

**Comment by zaki50** [comment]
fixed the indent issue.
:+1:

**Comment by nhachicha** [comment]
:+1:  squashed

**Comment by cmelchior** [comment]
retest this please

**Comment by nhachicha** [comment]
retest this please

**Comment by cmelchior** [comment]
Retest this please

**Comment by cmelchior** [comment]
retest this please

**Comment by nhachicha** [comment]
retest this please

**Comment by nhachicha** [comment]
retest this please

**Pull request merged**